examples/scene/bsn.rs

```diff
@@ -0,0 +1,197 @@
+// TODO: move these into actual tests and replace this with a more instructive user-facing example
+use bevy::{
+    prelude::*,
+    scene2::prelude::{Scene, *},
+};
+use bevy_scene2::SceneList;
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_systems(Startup, setup)
+        .add_systems(Update, (print::<Sprite>, print::<Transform>, print::<Team>))
+        .run();
+}
+
+fn setup(mut commands: Commands, assets: Res<AssetServer>) {
+    assets.load_scene("scene://base.bsn", base());
+    assets.load_scene("scene://transform_1000.bsn", transform_1000());
+    let top_level_handle = assets.load_scene("scene://top_level.bsn", top_level());
+    commands.spawn(ScenePatchInstance(top_level_handle));
+}
+
+fn top_level() -> impl Scene {
+    let a = 20usize;
+    let b = 1993usize;
+    bsn! {
+        #TopLevel
+        :"scene://base.bsn"
+        :x
+        Sprite { size: b }
+        Team::Green(10)
+        {transform_1337()}
+        Children [
+            Sprite { size: {4 + a}}
+        ]
+    }
+}
+
+fn base() -> impl Scene {
+    let sprites = (0..10usize)
+        .map(|i| bsn! {Sprite { size: {i} }})
+        .collect::<Vec<_>>();
+
+    bsn! {
+        Name("Base")
+        Sprite {
+            handle: "asset://branding/bevy_bird_dark.png",
+            size: 1,
+            nested: Nested {
+                handle: @"asset://hello.png"
+            }
+        }
+        Transform::from_translation(Vec3::new(1.0, 1.0, 1.0))
+        Team::Red {
+            x: 10,
+            y: Nested {
+                foo: 10
+            },
+        }
+        Gen::<usize> {
+            value: 10,
+        }
+        on(|event: On<Explode>| {
+        })
+        Foo(100, @"asset://branding/bevy_bird_dark.png")
+        [
+            (:sprite_big Sprite { size: 2 }),
+            :widget(bsn_list![Text::new("hi")]),
+            {sprites},
+        ]
+    }
+}
+
+fn sprite_big() -> impl Scene {
+    bsn! {
+        Sprite { size: 100000, handle: "asset://branding/icon.png" }
+    }
+}
+
+fn x() -> impl Scene {
+    bsn! {
+        :"scene://transform_1000.bsn"
+        Transform { translation: Vec3 { x: 11.0 } }
+    }
+}
+
+fn transform_1000() -> impl Scene {
+    bsn! {
+        Transform {
+            translation: Vec3 { x: 1000.0, y: 1000.0 }
+        }
+    }
+}
+
+fn transform_1337() -> impl Scene {
+    bsn! {
+        Transform {
+            translation: Vec3 { x: 1337.0 }
+        }
+    }
+}
+
+#[derive(Component, Debug, GetTemplate)]
+struct Sprite {
+    handle: Handle<Image>,
+    size: usize,
+    entity: Entity,
+    nested: Nested,
+}
+
+#[derive(Component, Debug, GetTemplate)]
+struct Gen<T: GetTemplate<Template: Default + Template<Output = T>>> {
+    size: usize,
+    value: T,
+}
+
+#[derive(Clone, Debug, GetTemplate)]
+struct Nested {
+    foo: usize,
+    #[template]
+    handle: Handle<Image>,
+}
+
+#[derive(Component, Clone, Debug, GetTemplate)]
```

Comment:
Why does GetTemplate need to be manually derived?

The GetTemplate traits feels... weird to me. Creating a template from a type seems like an internal detail to me. As a bevy user, I just want to think in terms of entities, assets, components, and helper functions that return impl Scene. Ideally I just derive component on my type and can automatically use it in scenes/templates.

Can we not just blanket drive GetTemplate, and hide it as an internal detail?

Reply:
That sadly goes directly against the point of GetTemplate.

The idea as I understand it, is that a Template is like a factory. And you can have several factories producing the same type of thing. So there's not a one-to-one correspondence between outputs and template types.

GetTemplate is there to express that sometimes there should be an obvious template for a type.
But I'm 99% sure this impl block is not valid, because it can and will create conflicting impls. (Even if it didn't create conflicting impls, rustc cannot prove it.)

Reply:
We could make a `ComponentTemplate<T: Component>(T)` type that implements GetTemplate, and then in the bsn! macro, it could automatically wrap components in ComponentTemplate.

Reply:
How would it know what Template logic to use for `T: Component`?